clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java

```diff
@@ -29,9 +32,9 @@
  */
 @InterfaceStability.Evolving
 public class DeleteConsumerGroupsResult {
-    private final Map<String, KafkaFuture<Void>> futures;
+    private final Map<CoordinatorKey, KafkaFutureImpl<Void>> futures;
 
-    DeleteConsumerGroupsResult(final Map<String, KafkaFuture<Void>> futures) {
+    DeleteConsumerGroupsResult(Map<CoordinatorKey, KafkaFutureImpl<Void>> futures) {
```

Comment:
@dajac @mimaison It looks like the reason we have to do `KafkaFutureImpl` here is because in the `SimpleAdminApiFuture` class, we have a class variable called `futures` which is of type ` private final Map<K, KafkaFutureImpl<V>>`. As a result, we return `KafkaFutureImpl` for the `all()`, `get()` and other methods. We could change all of this to use `KafkaFuture` instead, but this would require us to change the following methods from `protected` to `public`:

```java
    /**
     * If not already completed, sets the value returned by get() and related methods to the given
     * value.
     */
    protected abstract boolean complete(T newValue);

    /**
     * If not already completed, causes invocations of get() and related methods to throw the given
     * exception.
     */
    protected abstract boolean completeExceptionally(Throwable newException);
```

The question here would be should we change these `protected` methods to be `public` so we can maintain just returning a type of `KafkaFuture` or if it's okay to return `KafkaFutureImpl`. I think it could be worth changing `KafkaFuture` methods from `protected` to `public` so that we can maintain the invariant of always returning the type of `KafkaFuture`, but not sure if that would require getting some consensus on the mailing list about this.

Reply:
@skaundinya15 those methods in `KafkaFuture` are intentionally not `public` because a user receiving an instance should _never_ need to complete the future (they're always completed by the admin client). Making them `public` would thus make the API less type safe.

Reply:
@tombentley I see, thank you for the clarification. I guess in that case we have to stick with returning the `KafkaFutureImpl` in this situation then? Doesn't really seem to be a way around this otherwise.

Reply:
Yes that's exactly the reason why I made this change. I'll add the `final` modifier back though